docs/pages/pmd/userdocs/extending/writing_java_rules.md

```diff
@@ -0,0 +1,164 @@
+---
+title:  Writing a custom rule
+tags: [extending, userdocs]
+summary: "Learn how to write a custom rule for PMD"
+last_updated: July 3, 2016
+permalink: pmd_userdocs_extending_writing_java_rules.html
+author: Tom Copeland <tomcopeland@users.sourceforge.net>
+---
+
+{% jdoc_nspace :coremx core::lang.metrics %}
+{% jdoc_nspace :coreast core::lang.ast %}
+{% jdoc_nspace :jmx java::lang.java.metrics %}
+{% jdoc_nspace :jast java::lang.java.ast %}
+{% jdoc_nspace :jrule java::lang.java.rule %}
+
+{% include note.html content="TODO All that should be written in the Javadocs,
+not sure we even need a doc page. Would be simpler to maintain too" %}
+{% include warning.html content="WIP lots of stuff missing" %}
+
+This page covers the specifics of writing a rule in Java. The basic development
+process is very similar to the process for XPath rules, which is described in
+[Your First Rule](pmd_userdocs_extending_your_first_rule.html#rule-development-process).
+
+Basically, you open the designer, look at the structure of the AST, and refine
+your rule as you add test cases.
+
+In this page we'll talk about rules for the Java language, but the process is
+very similar for other languages.
+
+
+## Basics
+
+To write a rule in Java you'll have to:
+ 1. write a Java class that implements the interface {% jdoc core::Rule %}. Each
+language implementation provides a base rule class to ease your pain,
+e.g. {% jdoc jrule::AbstractJavaRule %}.
+ 2. compile this class, linking it to PMD APIs (eg using PMD as a maven dependency)
+ 3. bundle this into a JAR and add it to the execution classpath of PMD
```

Comment:
I some point, we might create a doc page about this -> https://github.com/pmd/pmd-examples